### PR TITLE
prov/efa: several bug fixes to FI_PEEK path

### PR DESCRIPTION
**Pull request by shijin-aws**
This PR contains 3 commits to fix bugs in the FI_PEEK related (FI_CLAIM, FI_DISCARD) code path.

This PR should be tested with a modified `rdm_tagged_peek` fabtest, which will be part of a separate PR to fabtests/functional (preparing now)

### CODE REVIEW

**Comment by shijin-aws** [comment]
Will push a new commit to add test after https://github.com/ofiwg/libfabric/pull/8770 is merged

File: prov/efa/src/rdm/rxr_msg.c [hunk 1]
```diff
@@ -1432,10 +1417,10 @@ ssize_t rxr_msg_trecvmsg(struct fid_ep *ep_fid, const struct fi_msg_tagged *tmsg
 	ep = container_of(ep_fid, struct rxr_ep, base_ep.util_ep.ep_fid.fid);
 
 	if (flags & FI_PEEK) {
-		ret = rxr_msg_peek_trecv(ep_fid, tmsg, flags);
+		ret = rxr_msg_peek_trecv(ep_fid, tmsg, flags | ep->base_ep.util_ep.rx_msg_flags);
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_msg.c:1420]: This looks fishy though. Why are we not respecting application flags? Maybe comment the reason here.

Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_msg.c:1420]: See the commit https://github.com/ofiwg/libfabric/commit/978ec0261d0a29b0e60a0de9ebb7b48f528e69ee

```
rxr_msg_recv (trecv, recvv, trecvv) should pass flags as
rxr_rx_flags(ep) to rxr_msg_generic_recv, which means it should
enable FI_COMPLETION in the flags as long as the prov_info->rx_attr.flags
support FI_COMPLETION. For rxr_msg_recvmsg (trecvmsg), it should
pass application flags |= util_ep.rx_msg_flags, which will have NO FI_COMPLETION
when application binds rx cq with FI_SELECTIVE_COMPLETION, and does not have
FI_COMPLETION in the flags of fi_recvmsg.
```

I can add comment in the code

Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_msg.c:1420]: This is how other providers (like shm) do as well

Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_msg.c:1432]: Updated

Comment (wenduwan) [reply prov/efa/src/rdm/rxr_msg.c:1432]: Thanks. This is better. It was odd that L1241 & L1302 were comment fine but not here.

**Review by wenduwan — APPROVED** [review]

**Review by wzamazon — APPROVED** [review]

**Comment by shijin-aws** [comment]
Latest push rebased on the latest libfabric main, and then added `rdm_tagged_peek` test to efa test suite.

**Review by wenduwan — APPROVED** [review]

**Comment by shijin-aws** [comment]
Pure efa change, ignoring Intel CI.

**Pull request merged**